crates/ty_python_semantic/src/types.rs

```diff
@@ -1029,7 +1034,48 @@ impl<'db> Type<'db> {
             return None;
         }
 
-        specialization
+        Some((class_type, specialization?))
+    }
+
+    /// Given a type variable `T` from the generic context of a class `C`:
+    /// - If `self` is a specialized instance of `C`, returns the type assigned to `T` on `self`.
+    /// - If `self` is a specialized instance of some class `A`, and `C` is a subclass of `A`
+    ///   such that the type variable `U` on `A` is specialized to `T`, returns the type
+    ///   assigned to `U` on `self`.
+    pub(crate) fn find_type_var_from(
+        self,
+        db: &'db dyn Db,
+        bound_typevar: BoundTypeVarInstance<'db>,
+        class: ClassLiteral<'db>,
+    ) -> Option<Type<'db>> {
+        if let Some(specialization) = self.specialization_of(db, class) {
+            return specialization.get(db, bound_typevar);
+        }
+
+        // TODO: We should use the constraint solver here to determine the type mappings for more
+        // complex subtyping relationships, e.g., `type[C[T]]` to `Callable[..., T]`, or unions
+        // containing multiple generic elements.
```

Comment:
you could possibly also mention protocols here, since although this PR tackles simple cases where a nominal type actually subclasses a protocol, there are also lots of cases where a nominal type is a subtype of a protocol without actually subclassing that protocol